### PR TITLE
Refactor cert gen, add more debug message and unit tests

### PR DESCRIPTION
**Pull request by edwardrf**

### CODE REVIEW

**Comment by edwardrf** [comment]
ref #400

File: src/pkg/cli/cert.go [hunk 1]
```diff
@@ -140,33 +171,62 @@ func waitForCNAME(ctx context.Context, domain, albDns string) error {
...
-				}
-				if doSpinner {
-					fmt.Print(spin.Next())
-				}
-			} else {
+			if checkDomainDNSReady(ctx, domain, albDns) {
 				return nil
 			}
+			if !msgShown {
+				term.Infof(" * Please setup CNAME record for %v", domain)
```
Comment (lionello) [inline src/pkg/cli/cert.go:178]: "CNAME or ALIAS"

File: src/pkg/cli/cert.go [hunk 2]
```diff
@@ -242,12 +327,30 @@ func getNSServers(ctx context.Context, domain string) ([]string, error) {
...
-		},
+func ipAddrsToIPs(ipAddrs []net.IPAddr) []net.IP {
+	ips := make([]net.IP, len(ipAddrs))
+	for i, ipAddr := range ipAddrs {
+		ips[i] = ipAddr.IP
+	}
+	return ips
+}
+
+func sameIPs(a, b []net.IP) bool {
```
Comment (lionello) [inline src/pkg/cli/cert.go:338]: test

File: src/pkg/cli/cert.go [hunk 3]
```diff
@@ -140,33 +171,62 @@ func waitForCNAME(ctx context.Context, domain, albDns string) error {
...
+func checkDomainDNSReady(ctx context.Context, domain, expectedCNAME string) bool {
+	expectedCNAME = strings.TrimSuffix(expectedCNAME, ".")
+	cname, err := getCNAMEInSync(ctx, domain)
+	term.Debugf(" - CNAME for %v is %v: %v", domain, cname, err)
+	// Ignore other types of DNS errors
+	if err == DNSNotInSyncError {
+		return false
+	}
+	cname = strings.TrimSuffix(cname, ".")
+	if strings.EqualFold(cname, expectedCNAME) {
```
Comment (lionello) [inline src/pkg/cli/cert.go:201]: allow more than 1 CNAME. In particular, allow public FQDN with and without --port

**Review by lionello — CHANGES_REQUESTED** [review]

File: src/pkg/cli/cert.go [hunk 4]
```diff
@@ -140,114 +176,188 @@ func waitForCNAME(ctx context.Context, domain, albDns string) error {
...
-					fmt.Printf("  %v  CNAME  %v\n", domain, strings.ToLower(albDns))
-					term.Infof(" * Waiting for CNAME record setup and DNS propagation...")
-					msgShown = true
-				}
-				if doSpinner {
-					fmt.Print(spin.Next())
-				}
-			} else {
+			if checkDomainDNSReady(ctx, domain, targets) {
+				sleep(ctx, 5*time.Second) // slight delay to ensure DNS to propagate
```
Comment (lionello) [inline src/pkg/cli/cert.go:180]: we have this in utils/pkg

**Review by lionello — APPROVED** [review]

**Pull request merged**